[Elixir] New Concept Exercise: lists (#1203)
exercism/v3: languages/elixir/exercises/concept/lists/.docs/introduction.md

> Lists are built-in to the Elixir language. They are considered a basic type, denoted by square brackets. Lists may be empty or hold any number of items. A list's contents are not required to have the same type:

Reviewer: ```suggestion
Lists are built-in to the Elixir language. They are considered a basic type, denoted by square brackets. Lists may be empty or hold any number of items of any type. For example:
```

```elixir
empty_list = []
one_item_list = [1]
two_item_list = [1, 2]
multiple_type_list = [1, :pi, 3.14]
```

> Elixir implements lists as a linked list, where each node stores the reference to the next list. The first item in the list is referred to as the _head_ and the remaining list of items is called the _tail_. We can use this notation in code:

Reviewer: ```suggestion
Elixir implements lists as linked lists, where each node stores the reference to the next item in the list. The first item in the list is referred to as the _head_ and the remaining list of items is called the _tail_. We can use this notation in code:
```

Reviewer: Conceptually, the introduction of lists as a list of items is probably easier and more similar to arrays in other languages.  If we want to point out that lists in Elixir are in fact hierarchical recursive data structures I think it might merit a bit more explanation.  The first time I realized that `[1, 2, 3]` really was something more like `[1, [2, [3]]]` it was a head trip.  The book "Elixir in Action" by Sasā Jurić has a good explanation of this, maybe we can dip into it?

Author: dip into it as another exercise? or a more detailed explanation?

Maybe there is a way to way to describe it at a high enough level in the introduction to get through the exercise, then explain more indepth in the after.md?

```elixir
# [1] represented in [head | tail] notation
[1 | []]

# [1, 2, 3 ]represented in [head | tail] notation
[1 | [2 | 3 | []]]
```

We can use _`[head | tail]`_ notation to prepend elements to a list:

```elixir
# Suppose
list = [2, 1]

[3, 2, 1] == [3 | list]
# => true
```

> There are several Elixir Kernel functions for working with lists, including `hd/1`, `tl/1`, and `length/1`. The `in/2` macro indicates whether an item is a member of a list, e.g.

Reviewer 2: I'm wondering it we need to list these functions in the introduction, or if we can refer to them in the `hints.md` file. Would that make sense?

> ```elixir

Reviewer 2: The `config.json` states that this exercise unlocks the `string-literals` concept. Is the below code meant as an introduction of what a string literal is?

Author: Yes, I want to introduce strings at a very high level without delving into string functions, because Elixir uses quite a bit of pattern matching concepts with those functions.  

Maybe this needs to briefly introduce strings as a discrete type without saying much more?

Reviewer 2: I'd say the description you use in the `after.md` would be just perfect here:

> [string][string] literals are a sequence of characters surrounded by double quotes

> There are also several Elixir Kernel functions for working with lists such as `hd/1`, `tl/1`, and `length/1`. One interesting macro function available is the `in/2` operator where we can see if an item is a member of a list.

Reviewer: ```suggestion
There are several Elixir Kernel functions for working with lists, including `hd/1`, `tl/1`, and `length/1`.  The `in/2` macro indicates whether an item is a member of a list, e.g.
```
